pkg/daemon/ceph/multus/utils.go

```diff
@@ -0,0 +1,286 @@
+package multus
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/coreos/pkg/capnslog"
+	"github.com/vishvananda/netlink"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+)
+
+var (
+	logger = capnslog.NewPackageLogger("github.com/rook/rook", "cephmultus")
+)
+
+const (
+	ifBase         = "mlink"
+	multusLinkName = "net1"
+	nsDir          = "/var/run/netns"
+)
+
+type multusNetStatus struct {
+	Name      string   `yaml:"name"`
+	Interface string   `yaml:"interface"`
+	Ips       []string `yaml:"ips"`
+}
+
+func GetMultusIP(pod corev1.Pod) (string, error) {
+	var multusIP string
+
+	if val, ok := pod.ObjectMeta.Annotations["k8s.v1.cni.cncf.io/networks-status"]; ok {
+		var multusStats []multusNetStatus
+
+		err := yaml.Unmarshal([]byte(val), &multusStats)
+		if err != nil {
+			return multusIP, err
+		}
+		for _, iface := range multusStats {
+			if iface.Interface == "net1" && len(iface.Ips) > 0 {
+				multusIP = iface.Ips[0]
+			}
+		}
+	} else {
+		return multusIP, errors.New("Multus Annotation not found")
+	}
+
+	return multusIP, nil
+}
+
+func determineHolderNS() (ns.NetNS, error) {
+	var holderNS ns.NetNS
+
+	holderIP, found := os.LookupEnv("HOLDERIP")
+	if !found {
+		return holderNS, errors.New("Environment variable HOLDERIP not set.")
+	}
+
+	log.Println("Finding the pod namespace handle")
+
+	nsFiles, err := ioutil.ReadDir(nsDir)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for _, nsFile := range nsFiles {
+		var foundNS bool
+
+		tmpNS, err := ns.GetNS(filepath.Join(nsDir, nsFile.Name()))
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		err = tmpNS.Do(func(ns ns.NetNS) error {
+			interfaces, err := net.Interfaces()
+			if err != nil {
+				return err
+			}
+
+			for _, iface := range interfaces {
+				link, err := netlink.LinkByName(iface.Name)
+				if err != nil {
+					return err
+				}
+				if link == nil {
+					return errors.New("Link not found")
+				}
+
+				addrs, err := netlink.AddrList(link, 0)
+				if err != nil {
+					return err
+				}
+
+				if len(addrs) < 1 {
+					continue
+				}
+
+				// Assuming that the needed address is first on the list.
+				if addrs[0].IP.String() == holderIP {
```

Comment:
a log message is probably needed here

Reply:
Actually I was able to rewrite it so this assumption is no longer needed.